src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,1518 @@
+
+#include "common/list.h"
+#include "common/_error.h"
+#include "common/xmltext_priv.h"
+#include "common/elements.h"
+#include "common/oscap_string.h"
+#include "common/util.h"
+
+#include "CPE/cpelang_priv.h"
+#include "CVSS/cvss_priv.h"
+#include "CVSS/public/cvss_score.h"
+
+#include "source/oscap_source_priv.h"
+#include "source/public/oscap_source.h"
+
+
+/***************************************************************************
+ ***************************************************************************
+ * CVRF structure definitions
+ */
+
+struct cvrf_index {
+	char *source_url;
+	char *index_file;
+	struct oscap_list *models;
+};
+OSCAP_ACCESSOR_STRING(cvrf_index, source_url)
+OSCAP_ACCESSOR_STRING(cvrf_index, index_file)
+OSCAP_IGETINS_GEN(cvrf_model, cvrf_index, models, model)
+OSCAP_ITERATOR_REMOVE_F(cvrf_model)
+
+struct cvrf_index *cvrf_index_new() {
+
+	struct cvrf_index *ret;
+
+	ret = oscap_alloc(sizeof(struct cvrf_index));
+	if (ret == NULL)
+		return NULL;
+
+	ret->source_url = NULL;
+	ret->index_file = NULL;
+	ret->models = oscap_list_new();
+
+	return ret;
+}
+
+void cvrf_index_free(struct cvrf_index *index) {
+
+	if (index == NULL)
+		return;
+
+	oscap_free(index->source_url);
+	oscap_free(index->index_file);
+	oscap_list_free(index->models, (oscap_destruct_func) cvrf_model_free);
+	oscap_free(index);
+}
+
+
+
+/***************************************************************************
+ * CVRF Model
+ * Top-level structure of the CVRF hierarchy
+ */
+struct cvrf_model {
+	char *doc_title;
+	char *doc_type;
+	struct cvrf_document *document;
+	struct cvrf_product_tree *tree;
+	struct oscap_list *vulnerabilities;	/* 1-n */
+};
+OSCAP_ACCESSOR_STRING(cvrf_model, doc_title)
+OSCAP_ACCESSOR_STRING(cvrf_model, doc_type)
+OSCAP_IGETINS_GEN(cvrf_vulnerability, cvrf_model, vulnerabilities, vulnerability)
+OSCAP_ITERATOR_REMOVE_F(cvrf_vulnerability)
+
+struct cvrf_product_tree *cvrf_model_get_product_tree(struct cvrf_model *model) {
+	return model->tree;
+}
+struct cvrf_document *cvrf_model_get_document(struct cvrf_model *model) {
+	return model->document;
+}
+const char *cvrf_model_get_identification(struct cvrf_model *model) {
+	struct cvrf_doc_tracking *tracking = cvrf_document_get_tracking(model->document);
+	return (cvrf_doc_tracking_get_tracking_id(tracking));
+}
+
+struct cvrf_model *cvrf_model_new() {
+
+	struct cvrf_model *ret;
+
+	ret = oscap_alloc(sizeof(struct cvrf_model));
+	if (ret == NULL)
+		return NULL;
+
+	ret->vulnerabilities = oscap_list_new();
+	ret->tree = cvrf_product_tree_new();
+	ret->doc_title = NULL;
+	ret->doc_type = NULL;
+	ret->document = cvrf_document_new();
+
+	return ret;
+}
+
+void cvrf_model_free(struct cvrf_model *cvrf) {
+
+	if (cvrf == NULL)
+		return;
+
+	oscap_free(cvrf->doc_title);
+	oscap_free(cvrf->doc_type);
+	cvrf_document_free(cvrf->document);
+	oscap_list_free(cvrf->vulnerabilities, (oscap_destruct_func) cvrf_vulnerability_free);
+	cvrf_product_tree_free(cvrf->tree);
```

Comment:
Seems to be in random order. I would expect reverse order as ctor. I could understand same order as ctor but this random order makes it harder to see that everything gets freed properly.